File: Medium/1930. Unique Length-3 Palindromic Subsequences/1930. Unique Length-3 Palindromic Subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        def count_unique_between(start, end, used):
            count = 0
            for j in range(start, end):
                idx = ord(s[j]) - ord('a')
                if not used[idx]:
                    used[idx] = True
                    count += 1
            for idx in range(26):
                used[idx] = False  # Resetting the values to False
            return count

        s_bytes = [ord(char) - ord('a') for char in s]
        left = [float('inf')] * 26
        right = [0] * 26
        ret = 0

        for i, char in enumerate(s_bytes):
            left[char] = min(left[char], i)
            right[char] = i

        used = [False] * 26
        for i in range(26):
            if left[i] != float('inf'):  # Check if left[i] is a valid index
                ret += count_unique_between(left[i] + 1, right[i], used)

        return ret

File: Medium/1930. Unique Length-3 Palindromic Subsequences/1930. Unique Length-3 Palindromic Subsequences.rs
impl Solution {
    pub fn count_palindromic_subsequence(s: String) -> i32 {
        let s = s.into_bytes();
        let mut ret = 0;

        let mut left = [usize::MAX; 26];
        let mut right = [0; 26];

        for (i, &b) in s.iter().enumerate() {
            let j = (b - b'a') as usize;
            left[j] = left[j].min(i);
            right[j] = i;
        }

        let mut used = [false; 26];
        for i in 0..26 {
            ret += count_unique_between(&s, left[i] + 1, right[i], &mut used);
        }

        ret
    }
}

fn count_unique_between(s: &[u8], start: usize, end: usize, used: &mut [bool; 26]) -> i32 {
    let mut count = 0;
    for j in start..end {
        let idx = (s[j] - b'a') as usize;
        if !used[idx] {
            used[idx] = true;
            count += 1;
        }
    }
    used.fill(false);
    count
}
